### ISSUE TITLE
Test for POP ES instruction with different operand size

### ISSUE THREAD

**Issue by barotto** [barotto/test386.asm#4]
From VOGONS thread https://www.vogons.org/viewtopic.php?f=9&t=60095

> Another code bug revealed with the POP ES instruction: it wasn't using the operand size to perform the 16-bit POP. So it was always increasing/decreasing (E)SP with 2 instead of the proper 4.

Test for this.

**Comment by superfury** [comment]
You might as well make it a full push/pop test with different operand sizes and src/destinations? So all basic segment register push/pop, opcode 8F pop, opcode ff push, 80186+ immediate pushes(push imm8/16/32), off the top of my head.

**Comment by barotto** [comment]
I know, some of those instructions maybe used but none of them are explicity tested, yet.
If an instruction is "officially" tested, it is marked as such in in the opcodes list (intel-opcodes.ods). That list also discriminates for operand size, address size, and stack size.

**Comment by barotto** [comment]
PUSHA/POPA are heavily used for calls, that doesn't mean they are bug free though. They have to be properly tested for corner cases.
This is the count of all the instructions currently used: [icnt.txt](https://github.com/barotto/test386.asm/files/2050693/icnt.txt)

> Even most basic pretty important instructions are left unchecked?

Yep.

> Why isn't INC/DEC(r/m16/32)/CMP/TEST tested with test 0xEE?

Because there are only 24 hours in a day, usually :)

**Comment by barotto** [comment]
Added:
0E PUSH CS
1E PUSH DS
1F POP DS
16 PUSH SS
17 POP SS
06 PUSH ES
07 POP ES
0FA0 PUSH FS
0FA1 POP FS
0FA8 PUSH GS
0FA9 POP GS

**Comment by superfury** [comment]
So, that still leaves:

60 PUSHA(D)
61 POPA(D)
68 PUSH imm16/32
6A PUSH imm8
8F /0 POP r/m16/32
9C PUSHF(D)
9D POPF(D)
FF /6 PUSH r/m16/32

As well as those INC/DEC(16&32 bits variants)/CMP/TEST on test EE(where, CMP=SUB and TEST=AND on a logical flag level(unaffected registers other than flags). INC/DEC is about the same as ADD/SUB with 1, but not affecting the carry flag).

**Comment by barotto** [comment]
Can you test an older version of your emu to see if that POP ES bug you talked about at vogons can now be detected with test386?

>  self-modifying code 
I think that would be a bit overkill... I'll probably test with some pattern.

**Comment by superfury** [comment]
I've just tested the current testsuite against a bit older commit of UniPCemu, before the POP ES fix. It will properly error out during the POST 0A tests.

**Comment by superfury** [comment]
Don't you mean 8F /0? 8F /1-7 are #UDs.

**Comment by barotto** [comment]
Yes you're right, 8F/0, cut'n'paste error...

**Comment by barotto** [comment]
Added:
60 PUSHA(D)
61 POPA(D)

**Comment by barotto** [comment]
Added:
68 PUSH imm16/32
6A PUSH imm8

**Comment by superfury** [comment]
They all still check out. That just leaves PUSHF(D) and POPF(D) instructions. All other PUSH/POP operations are tested now.

**Comment by barotto** [comment]
> They all still check out.

Is this a good or a bad thing? Do you expect any test to fail?

**Comment by superfury** [comment]
Well, I know some software not running properly, most of them not even running in protected mode(yet). Like California Games hanging on the stage select screen(any CPU, 8086+ problem), Windows 95 crashing into cleared RAM(containing 0000h opcodes), Day of The Tentacle(tentacle.exe) not seeing the command line and giving me a unexpected 'by 0\n' on the command line, before showing help on parameters then crashing with a NULL pointer load error. Windows 3.0 simply hangs with recursive protection faults in protected mode after IRET from the kernel.

**Comment by superfury** [comment]
Odd that with the latest improvements, some of those mentioned software(Windows 95 setup, Jazz Jackrabbit) tries to execute Protected-mode only instructions in real mode now(ARPL for Windows 95 setup, LLDT for Jazz Jackrabbit)?

**Comment by barotto** [comment]
ARPL and LLDT are protected mode only and #UD would be thrown if a program would try to execute them in real mode. Sounds like some jump not taken somewhere way before those instructions.

**Comment by superfury** [comment]
Well, that's the strange part: one of those two(DOTT) is made for the 80286. So it just supports 8086 jumps. And those should all be working correctly(manually tested each one of them using simple breakpoints(in Visual Studio) on those instructions. No errors were observed.

**Comment by barotto** [comment]
Well, maybe the JMPs themselves are good but the code that decided to take or not those jumps is wrongly executed. The problem could actually be a thousand cycles before...
Do you use a prefetch queue? I once fixed a bug that made the prefetcher read the wrong code memory area.

**Comment by superfury** [comment]
Just checked all x86 jumps and conditional jumps. Each and every one should be working properly as far as I can see?

https://bitbucket.org/superfury/unipcemu/src/a32836f0af33adbbc690098ed72a52dcad23f18c/UniPCemu/cpu/?at=master

Look at opcodes*.c for all opcode implementations.

Look for CPU_jmprel for relative jumps and calls, CPU_jmpabs for absolute ones. "segmentWritten(CPU_SEGMENT_CS"(and destEIP load before that for it's EIP address to load) for the far jumps and calls.

I can't see any errors in those instructions processing.

Perhaps an error in the flags itself could also be a cause, but errors even happen on the 8086, which just pushes and pops them without modification of said data(raw access).

**Comment by superfury** [comment]
Just tried running Windows 3.0 in it's default mode(no parameters). I see it page faulting twice(no other faults), then returning to the MS-DOS 5.0 prompt with the Windows logo screen behind it.

**Comment by superfury** [comment]
I've verified(during execution with breakpoints on the flags in Visual C++) that most conditional jump and loop instructions work as documented. Both used breakpoints on condition match and non match after it being matched. Jumps are taken when required and not taken when required not to.

The only cases left untested are JO(overflow flag not set), JNO and JP(parity flag not set). All other jxx instructions work correctly.

**Comment by superfury** [comment]
Yes, a PIQ is emulated, but it should be functioning normally afaik.

**Comment by superfury** [comment]
Also, the BIU must be working properly, otherwise it can't even boot past the BIOS, since any jump would fail right away because of a wrong base address/offset.

**Comment by superfury** [comment]
Just tried running the AT BIOS, which confirms proper execution of the JP and JO conditional jumps. The only one left untested is JNO, both taken and not taken.
